core/src/main/java/com/microsoft/alm/auth/oauth/OAuth2Authenticator.java

```diff
@@ -154,7 +159,59 @@ public TokenPair getOAuth2TokenPair(final PromptBehavior promptBehavior) {
 
         final String key = getKey(APP_VSSPS_VISUALSTUDIO);
 
-        SecretRetriever secretRetriever = new SecretRetriever() {
+        final SecretRetriever<TokenPair> secretRetriever = new SecretRetriever<TokenPair>() {
+
+            private boolean validateAccessToken(final Token accessToken, final URI validationEndpoint) {
+                final HttpClient client = new HttpClient(Global.getUserAgent());
+                accessToken.contributeHeader(client.Headers);
+                try {
+                    final HttpURLConnection response = client.get(validationEndpoint);
+
+                    return response.getResponseCode() == HttpURLConnection.HTTP_OK;
+                } catch (IOException e) {
+                    logger.debug("Validation failed with IOException.", e);
+                }
+
+                return false;
+            }
+
+            @Override
+            protected boolean tryGetValidated(final TokenPair tokenPair, final AtomicReference<TokenPair> holder) {
+                Debug.Assert(tokenPair != null, "TokenPair is null");
+                Debug.Assert(holder != null, "Holder is null");
+
+                final URI validationEndpoint = URI.create(VALIDATION_ENDPOINT);
+                boolean valid = false;
+
+                if (tokenPair.AccessToken != null) {
+                    logger.debug("Validating stored OAuth2 Access Token...");
+                    valid = validateAccessToken(tokenPair.AccessToken, validationEndpoint);
+                }
+
+                if (!valid && tokenPair.RefreshToken != null) {
+                    logger.debug("OAuth2 Access Token is not valid, and we have a refresh token, try refreshing...");
+
+                    final TokenPair renewedTokenPair =
+                            getAzureAuthority().acquireTokenByRefreshToken(clientId, resource, tokenPair.RefreshToken);
+
+                    // Today after we refresh oauth2 token with the refresh token, we are only getting back an
+                    // Access Token -- the refresh token is null.  Although we could use this Access Token for another
+                    // hour, this contradicts with the assumption of this library (Both AccessToken and
+                    // RefreshToken fields of the TokenPair class are non-null).
+                    // TODO: investigate next sprint, after 7/28
+                    if (renewedTokenPair != null
+                            && renewedTokenPair.AccessToken.Value != null
+                            && renewedTokenPair.RefreshToken.Value != null) {
+                        logger.debug("OAuth2 Access Token refreshed successfully.");
+                        valid = true;
+                        holder.set(renewedTokenPair);
+                    }
+                }
+
+                logger.debug("OAuth2 Access Token is {}.", valid ? "valid" : "invalid.");
+                return valid;
```

Comment:
We do not log about "expiration" specifically since we don't save that information.  However if the AccessToken is not valid, and we do have a refresh token, we log the fact we will try to refresh.  By omission of that log line and if validation failed, we can deduce we don't have a refresh token.

This piece of code will be revisited soon.